### PR TITLE
Added scope validation

### PR DESCRIPTION
**Pull request by ruks**
Added scope validation

### CODE REVIEW

**Comment by harsha89** [comment]
There is a feature which allow to add scopes without role validation. Are we checking that propert during this implementation?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 1]
```diff
@@ -3376,4 +3380,39 @@ public String getMediationNameFromConfig(String config) {
...
+                String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
+                int tenantId = APIUtil.getTenantIdFromTenantDomain(tenantDomain);
+                APIProvider apiProvider = RestApiUtil.getProvider(username);
+
+                if (apiProvider.isScopeKeyAssigned(apiId, scope.getName(), tenantId)) {
+                    RestApiUtil
+                            .handleBadRequest("Scope " + scope.getName() + " is already assigned by another API", log);
+                }
+            }
+            if (scope.getDescription() == null || scope.getDescription().trim().isEmpty()) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: Can't we use stringutils here

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 2]
```diff
@@ -3376,4 +3380,39 @@ public String getMediationNameFromConfig(String config) {
...
+                RestApiUtil.handleBadRequest("Scope cannot have empty description", log);
+            }
+            if (scope.getRoles() != null) {
+                for (String aRole : scope.getRoles().split(",")) {
+                    boolean isValidRole = APIUtil.isRoleNameExist(apiId.getProviderName(), aRole);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: why we need to pass provider name to check role existents ? Is it due to get the tenant domain?

Comment (ruks) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: yes. Existing APIUtil#isRoleNameExist() require tenant ID. it is extracted from username.

**Comment by ruks** [comment]
> There is a feature which allow to add scopes without role validation. Are we checking that propert during this implementation?
Yes. APIUtil.isRoleNameExist() check that system property internally.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 1 above)
Comment (ruks) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: yes. Fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 3]
```diff
@@ -3376,4 +3380,39 @@ public String getMediationNameFromConfig(String config) {
...
+                int tenantId = APIUtil.getTenantIdFromTenantDomain(tenantDomain);
+                APIProvider apiProvider = RestApiUtil.getProvider(username);
+
+                if (apiProvider.isScopeKeyAssigned(apiId, scope.getName(), tenantId)) {
+                    RestApiUtil
+                            .handleBadRequest("Scope " + scope.getName() + " is already assigned by another API", log);
+                }
+            }
+            if (StringUtils.isBlank(scope.getDescription())) {
+                RestApiUtil.handleBadRequest("Scope cannot have empty description", log);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: This is not nexesary; Why the description should mandate?

**Pull request merged**